### PR TITLE
[No QA] Onyx type core data

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details
Iteration on https://github.com/Expensify/App/pull/19966. PR was reviewed [here](https://github.com/blazejkustra/App/pull/1) by internal engs by referencing backend code.

TODO:

- [x]  Type the core Onyx data shapes
- [x]  Install typed version of react-native-onyx. Blocked by https://github.com/Expensify/react-native-onyx/pull/267
- [x]  Overwrite react-native-onyx in global.d.ts
- [x]  Expensify engineers check the typed format by referencing backend code cc @hayata-suenaga

### Fixed Issues

$ https://github.com/Expensify/App/issues/16665
PROPOSAL: N/A

### Tests
N/A

### Offline tests
N/A

### QA Steps
N/A

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
cc @neil-marcellini

File: package.json
Comment (blazejkustra) [inline package.json:225]: Removed as discussed [here](https://swmansion.slack.com/archives/C0593T50UHG/p1691172993356179?thread_ts=1691158410.374569&cid=C0593T50UHG)

File: package.json
Comment (blazejkustra) [inline package.json:247]: Updated as described [here](https://swmansion.slack.com/archives/C0593T50UHG/p1691158410374569)

**Review by hayata-suenaga — DISMISSED** [review]

**Comment by s77rt** [comment]
@blazejkustra Is this ready for review?

**Comment by hayata-suenaga** [comment]
@s77rt this PR has been reviewed by bunch of internal engineers in [this PR](https://github.com/blazejkustra/App/pull/1). (we had to create this one because that PR's target branch is a brunch of the forked repo).

So if you can test generally the build succeeds and app works, that's all needed here 🙇

**Comment by hayata-suenaga** [comment]
@s77rt and for the context, we're migrating App to TypeScript and this PR defines types for data stored on Onyx.

**Comment by blazejkustra** [comment]
> Just to confirm, this one is on hold until the https://github.com/Expensify/react-native-onyx/pull/267 is merged to main and the npm package is updated right?

That's correct @hayata-suenaga

File: src/types/onyx/Report.ts
Comment (s77rt) [inline src/types/onyx/Report.ts:49]: I think it's always an array of strings (logins)

File: src/types/onyx/Report.ts
Comment (s77rt) [inline src/types/onyx/Report.ts:72]:
```suggestion
    parentReportID?: string;
```

This is supposed to be a string https://expensify.slack.com/archives/C01GTK53T8Q/p1691167905181059

File: src/types/onyx/Report.ts
Comment (s77rt) [inline src/types/onyx/Report.ts:81]:
```suggestion
    lastActorAccountID?: number;
```

File: src/types/onyx/ReportAction.ts
Comment (s77rt) [inline src/types/onyx/ReportAction.ts:12]: Those are defined in `OriginalMessage`

File: src/types/onyx/ReportAction.ts
Comment (s77rt) [inline src/types/onyx/ReportAction.ts:82]:
```suggestion
    childReportID?: string;
```
Also supposed to be a string (related to parentReportID)

File: src/types/onyx/Task.ts
Comment (s77rt) [inline src/types/onyx/Task.ts:21]:
```suggestion
    assigneeAccountID?: number;
```

File: src/types/onyx/Transaction.ts
Comment (s77rt) [inline src/types/onyx/Transaction.ts:5]: NAB, do we really need a type for Comment?

File: src/types/onyx/WalletTransfer.ts
Comment (s77rt) [inline src/types/onyx/WalletTransfer.ts:6]: I think this may be both a number or a string based on `src/pages/settings/Payments/walletTransferPropTypes.js`

**Comment by s77rt** [comment]
Thanks @hayata-suenaga I have just added some comments regarding types. Other than that I think I will just approve the PR. Excited for TS migration :heart:

**Review by neil-marcellini — DISMISSED** [review]
The PR this was created from has been very thoroughly reviewed so this is good to go as well 👍

File: src/types/onyx/Transaction.ts
Comment (blazejkustra) [reply src/types/onyx/Transaction.ts:5]: From TS guidelines (TS_STYLE.md): 

> If T isn't a primitive type, create a separate type for the T object structure of your prop and use it.

File: src/types/onyx/Task.ts
Comment (blazejkustra) [reply src/types/onyx/Task.ts:21]: Thanks, adjusted 👍

**Comment by blazejkustra** [comment]
I resolved the conflicts and adjusted the PR after @s77rt initial review. I also installed newest version of typed `react-native-onyx` package. @s77rt please have another look! I think we need a PR reviewer checklist to pass all checks.

**Comment by blazejkustra** [comment]
There were some big changes in Onyx recently (mentioned [here](https://github.com/Expensify/App/pull/24212#issuecomment-1668259211)), when I updated Onyx to the latest version it caused some test failures.

The Onyx version is bumped in this [PR](https://github.com/Expensify/App/pull/24041) & tests are fixed. So we are pretty much blocked until that PR is merged. 

cc @hayata-suenaga @fabioh8010

File: src/types/onyx/Transaction.ts
Comment (s77rt) [reply src/types/onyx/Transaction.ts:5]: I'm not sure this is the case here. Comment is of type string. This is a primitive type.

**Review by s77rt — APPROVED** [review]
Just a minor note https://github.com/Expensify/App/pull/24054#discussion_r1287373431. Other than that LGTM! :rocket:

**Review by neil-marcellini — DISMISSED** [review]

**Comment by hayata-suenaga** [comment]
will review this tonight

File: src/types/onyx/Transaction.ts
Comment (hayata-suenaga) [reply src/types/onyx/Transaction.ts:5]: @s77rt thank you for very detailed review of this PR 😊

`comment` here is actually the following type

```ts
{
  ...
  comment: {
    comment: string;
  }
  ...
}
```

so based on our guideline, we have to do

```ts
type Comment = {
  commment: string;
}

{
  ...
  comment: Comment
  ...
}
```

**Review by hayata-suenaga — DISMISSED** [review]
🥳 🥳 🥳 🥳 🥳

File: src/types/onyx/Transaction.ts
Comment (s77rt) [reply src/types/onyx/Transaction.ts:5]: Ah I see that makes sense! In that case shouldn't we add `originalTransactionID` and `source` as they are a part of the comment object (both are strings).

File: package.json
Comment (hayata-suenaga) [inline package.json:122]: we actually have to change this to the actual Onyx version after [this PR](https://github.com/Expensify/react-native-onyx/pull/267) is merged and the version is bumped and the new version is published to npm.

but yes, we have to wait for [this PR](https://github.com/Expensify/App/pull/24041) before doing that

**Comment by hayata-suenaga** [comment]
also we apparently have a merge conflict @blazejkustra when you come back, could you solve it please 🙇

**Comment by hayata-suenaga** [comment]
I gonna approve the PR once the blocking PR has been merged main and main is merged to this PR.

Then the Jest tests should pass

**Review by hayata-suenaga — DISMISSED** [review]

**Review by neil-marcellini — DISMISSED** [review]
@blazejkustra looking good but we have some merge conflicts again.

**Review by hayata-suenaga — APPROVED** [review]

File: src/types/onyx/ReportAction.ts
Comment (roryabraham) [inline src/types/onyx/ReportAction.ts:47]: I know this was a giant PR, but shouldn't `reportActionID` be required + non-null?

Comment (fabioh8010) [reply src/types/onyx/ReportAction.ts:47]: @blazejkustra

Comment (blazejkustra) [reply src/types/onyx/ReportAction.ts:47]: I mimicked the proptypes from [here](https://github.com/software-mansion-labs/expensify-app-fork/blob/7fe7fe5c9e297e500d087637472713c491f222cd/src/pages/home/report/reportActionPropTypes.js#L6). There were some usages of `reportActionID` with a default value, that's probably why it's optional.

The thing is that onyx core data types aren't final and we'll edit these as we continue the migration to Typescript (it wasn't feasible to catch all the usages 😅)

File: src/types/onyx/Task.ts
Comment (tgolen) [inline src/types/onyx/Task.ts:10]: @blazejkustra This field does not exist in the API or the database from what I can tell. What should we do with this TODO?

Also, please don't add TODOs to the code. Instead, open a GH so that there is a very actionable way to track what needs to happen.

Comment (blazejkustra) [reply src/types/onyx/Task.ts:10]: Ahh sorry, this was quite a big PR and I forgot about this TODO, so this kind of proves your point 👍 

Could you create an issue to remove `shareDestination` from the E/App? It is used in various places, and if `shareDestination` doesn't come from API/database we should be able to safely remove some of the logic:
image

Comment (hayata-suenaga) [reply src/types/onyx/Task.ts:10]: [Created the issue here](https://github.com/Expensify/App/issues/29402)

**Pull request merged**
